File: HanSeokhyeon/TargetNumber.java
/*
타겟 넘버

-1+1+1+1+1 = 3
+1-1+1+1+1 = 3
+1+1-1+1+1 = 3
+1+1+1-1+1 = 3
+1+1+1+1-1 = 3

DFS로 depth만큼 가보고 결과값이 target과 같으면 answer++
 */

public class TargetNumber {

    private int answer = 0;
    private int[] numbers;
    private int target;

    public static void main(String[] args) {
        TargetNumber targetNumber = new TargetNumber();
//        int solution = targetNumber.solution(new int[]{1, 1, 1, 1, 1}, 3);
        int solution = targetNumber.solution(new int[]{4, 1, 2, 1}, 4);
        System.out.println("solution = " + solution);
    }

    public int solution(int[] numbers, int target) {
        this.numbers = numbers;
        this.target = target;
        dfs(0, 0);
        return answer;
    }

    private void dfs(int depth, int value) {
        if (depth == numbers.length) {
            if (value == target) {
                answer++;
            }
            return;
        }
        dfs(depth+1, value+numbers[depth]);
        dfs(depth+1, value-numbers[depth]);
    }
}

File: MinJunKim/targetnumber.js
내가 푼거 아님
재귀가 되도록 구성을 해야할 듯 하였는데, dfs 문제인지 아직 모르는 수준...

function solution(numbers, target) {
    let answer = 0;
    
    dfs(0, 0);
    
    function dfs(index, sum) {
        if(index === numbers.length) {
            if (sum === target) {
                answer++;
             }
            return;
        }
        
        dfs(index + 1, sum + numbers[index]);
        dfs(index + 1, sum - numbers[index]);
    }
    
    return answer;
}
